lib/lightning/kafka_triggers/pipeline.ex

```diff
@@ -126,13 +127,25 @@ defmodule Lightning.KafkaTriggers.Pipeline do
   def handle_failed(messages, context) do
     messages
     |> Enum.each(fn message ->
+      publish_failure_event(message, context)
       notify_sentry(message, context)
       create_log_entry(message, context)
     end)
 
     messages
   end
 
+  defp publish_failure_event(
+         %{status: {:failed, :persistence}} = message,
+         %{trigger_id: trigger_id}
+       ) do
+    %{metadata: %{ts: timestamp}} = message
+
+    Events.kafka_trigger_persistence_failure(trigger_id, timestamp)
```

Comment:
If a persistent rewind won't be used for now, we should at least process the failure synchronously. If the persistence failure event is handled asynchronously, a message with higher timestamp than the `timestamp` of the failure might be processed twice on this timeline:
```
1. Broadway.Message.failed(:persistence) for timestamp t0
2. Event is published for timestamp t0
3. Broadway pipeline receives next message with timestamp t1
4. A workorder is created for t1
5. Event for timestamp t0 is proccessed and pipeline is reset to timestamp t0
6. Pipeline consumes timestamp t0 but creates another workorder for the message with timestamp t1.

Reply:
So summing it up there would be two concerns:
i) The persistence/assurance that the failures will be handled
ii) When the failure is handled (if duplicated work orders need to be prevented).

Reply:
@jyeshe wrt (ii) we already have a deduplication mechanism in place for incoming messages as this is what is required to consume messages from Kafka (the Kafka architecture means that there is always a chance that it will serve up a duplicate message). 

The message is stored in the Db and retained for a period of time before being expunged, so in your timeline above, an entry would be made for `t1` at step (4) in the `trigger_kafka_message_records` table. As long as `t1` arrives for the second time, within the period that `t1` is resident in `trigger_kafka_message_records`, step (6) will ignore the message.

Wrt (i) I think I am failing to properly understand what you are suggesting: If we are to reset the trigger pipeline with the new timestamp synchronously, would that not mean that the pipeline would first need to stop itself? At which point, what would be responsible for starting it?

Reply:
@jyeshe Apologies, I conflated the synchronous processing with the persistent rewind. Please ignore what I asked wrt (i). For (i) are you suggesting that we persist a 'rewind' flag to the DB that is then picked up by the Trigger on restart, or do you have another persistence mechanism in mind?

Reply:
Hey @rorymckinley I wouldn't say it's from Kafka architecture but a retry setup on their producer side once Lightning/Broadway already commits the offsets which are the two conditions for `at-most-once` policy.

wrt (ii) I see the `trigger_kafka_message_recrods` has `(trigger_id, topic_partition_offset)` as a primary key but this deduplication would cause `track_message` to fail.

The problem with (i) is that if the app crashes or there is a k8s downscale or the cluster is stopped for maintenance between the moment the `Pipeline.publish_failure_event` is called and `KafkaTriggers.EventListener` processes the event, the `Events.KafkaTriggerPersistenceFailure` is lost and the offset is not rewinded. If the event is broadcasted on Lightning PubSub there is a chance another pod replica to process it but on crash app or cluster stop there would be still a risk of not processing the `Events.KafkaTriggerPersistenceFailure`.